882 Implement component diagnostic interface

lifflander opened the pull request:
Fixes #882 

- [x] write diagnostic interface tests
- [x] add diagnostics for `AsyncEvent`
- [x] add diagnostics for scheduler
- [x] add outputting of diagnostics at finalization/exigent exit
- [x] compile-time feature for diagnostics to disable
- [x] runtime argument to enable/disable diagnostics
<img width="1499" alt="Screen Shot 2020-07-01 at 4 04 59 PM" src="https://user-images.githubusercontent.com/3324465/86299337-dcbb6f00-bbb4-11ea-8d8f-241ad21d30b5.png">

lifflander commented on src/vt/configs/types/types_type.h:
```diff
@@ -106,6 +106,8 @@ using GroupType               = uint64_t;
 /// Used for hold a phase for load balancing
 using PhaseType               = uint64_t;
+/// Used for hold a sub-phase for load balancing
+using SubphaseType            = uint16_t;
```
This refactor was required to get the code compiling due to some circular includes with the new PR that I introduced later on.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented on src/vt/scheduler/scheduler.cc:
```diff
@@ -67,11 +67,98 @@ namespace vt { namespace sched {
+  using timing::Timing;
+
+  // Triggers to get the in-scheduler-loop time added to diagnostics
+  registerTrigger(
```
I'm concerned about how expensive this might be to do with triggers. Registering triggers for itself is definitely the cleanest way to do this, but we could accumulate these times up out-of-line to make this cheaper.

PhilMiller commented on src/vt/messaging/active.cc:
```diff
@@ -74,6 +74,97 @@ ActiveMessenger::ActiveMessenger()
+
+  // Number of broadcast messages that hit this node
+  registerDiagnostic<int64_t>(
+    "bcasts", "active message broadcasts", UpdateType::Sum
```
Should this be `bcasts_received`, to be unambiguous vs number sent

lifflander replied:
Fixed

PhilMiller commented on src/vt/runtime/component/diagnostic_enum_format.cc:
```diff
@@ -0,0 +1,115 @@
+  return true;
+}
+
+std::string diagnosticUnitTypeString(DiagnosticUnit unit) {
```
Would it make sense to add a `DiagnosticUnit::Occurrences` for the things that are counted, rather than leaving them as unknown?

PhilMiller commented on src/vt/runtime/component/diagnostic_value.h:
```diff
@@ -0,0 +1,325 @@
+  }
+
+  /**
+   * \internal \brief Get min value (use after reduction)
```
Copy-paste error: max vs min

lifflander replied:
Fixed

PhilMiller commented on src/vt/runtime/component/diagnostic_value.h:
```diff
@@ -0,0 +1,325 @@
+      value_.update(std::max(value_.getRawValue(), val));
+      break;
+    case DiagnosticUpdate::Avg:
+      value_.incrementN();
```
Comment the fallthrough here, or preferably just write out the one duplicated line.

lifflander replied:
Added

PhilMiller commented:
This looks like a really great start. Once whatever bits and pieces are cleaned up, it may be more sensible to merge with just the AM diagnostics, and follow on with other components in separate PRs. It would certainly reduce the risk of merge conflicts developing as things progress.

pnstickne commented on src/vt/runtime/component/diagnostic_value_format.h:
```diff
@@ -0,0 +1,256 @@
+    case DiagnosticUnit::Bytes: {
+      // Start with the largest unit, testing if its appropriate for the value,
+      // and if not, downgrade it until we find one that works
+      auto multiplier = static_cast<int8_t>(MemoryUnitEnum::Gigabytes);
```
Nit: `GigiBytes` (or `Gigibytes`).

lifflander replied:
Good point. These are all x1024 unit increments.

pnstickne commented on src/vt/event/event.cc:
```diff
@@ -298,11 +324,27 @@ void AsyncEvent::testEventsTrigger(int const& num_events) {
   auto& cont = polling_event_container_;
+
+  if (cont.size() > 0) {
+    updateDiagnostic<int64_t>("avg_events_size", cont.size());
```
Using declare/update approach feel like it might allow for better optimization in the future. (As done for events, performance counters, etc..)

     Diag<..> avgEventSize = getDiagnostic<int64_t>("avg_events_size");
     updateDiag(avgEventSize, config.size());
     // or
     avgEventSize.add(count.size());

lifflander replied:
Ok, I've changed the implementation and interface with a declare/use approach saved in a variable.

pnstickne commented on src/vt/event/event.cc:
```diff
@@ -298,11 +324,27 @@ void AsyncEvent::testEventsTrigger(int const& num_events) {
+      auto duration = timing::Timing::getCurrentTime() - event->getCreateTime();
+      updateDiagnostic<double>("avg_mpi_event_wait", duration);
+      updateDiagnostic<double>("max_mpi_event_wait", duration);
+#     endif
```
Similarly,

     waitDuration.time(event->getCreateTime(), timing::Timing::getCurrentTime());

This case is a little odd because of separate "avg" and "max". However, in statsd (and to a lesser extent windows performance counters), these are the result of computations not an artifact of being added to different data sinks (of the same kind).

lifflander replied:
Done

pnstickne commented on src/vt/event/event.cc:
```diff
@@ -298,11 +324,27 @@ void AsyncEvent::testEventsTrigger(int const& num_events) {
+      updateDiagnostic<double>("avg_mpi_event_wait", duration);
+      updateDiagnostic<double>("max_mpi_event_wait", duration);
+#     endif
+
```
Also, one thing that early stastd-style implementation failed at was the ability to collect axis information.. this has been fixed it later implementations, so keeping that as a possibility..

lifflander replied:
Addressed

pnstickne commented:
Pretty nifty. I think any stats is better than no stats, however the true value from such metrics is to use Y-plots against X in buckets of some fixed interval. The most _useful_ end goal would be to have this data plotted _against_ the current projections visualization.

For "industry" code this interval is often 1 minute (although _this time is arbitrary_, it's the concept that is important) over which time the aggregates for each metric will be computed. This is usually sum/min/max/avg (sum can also span multiple buckets) - so basically what is displayed at the end, only collected (and usual reset) in some interval.

pnstickne commented:
Not sure if can "require MPI 3.0" for this extra, then https://computing.llnl.gov/projects/mpi_t MPI-T might be useful to interface with. Even if not, it might be a good example.

pnstickne commented:
Here is the "basis" for most current stats metrics - https://code.flickr.net/2008/10/27/counting-timing/

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  2
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/diagnostic_meter.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  2
- src/vt/runtime/component/diagnostic_meter.h  2
- tests/unit/utils/test_histogram_approx.cc  4
- src/vt/runtime/component/diagnostic.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic_meter.h  2
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/diagnostic_value_format.h  2
- tests/unit/utils/test_histogram_approx.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/diagnostic_meter.h  2
- src/vt/runtime/component/diagnostic_value_format.h  2
- tests/unit/utils/test_histogram_approx.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic_meter.h  2
- src/vt/runtime/component/diagnostic_value_format.h  2
- src/vt/runtime/component/diagnostic.impl.h  2
- tests/unit/utils/test_histogram_approx.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
@PhilMiller @pnstickne Please take a look at the new interface. I've taken your suggestions and implemented counters, gauges, and timers. Now, the user automatically gets the min/max/avg/etc for gauges and timers and we have higher-level info like the start and end time for a timer.

I've also implemented a streaming histogram algorithm with a bounded size based on this paper:
 http://www.jmlr.org/papers/volume11/ben-haim10a/ben-haim10a.pdf

- I'm using that implementation to build a histogram over ranks/nodes during the reduction, while keeping the amount of data constant so its scalable with P. 
- I plan to use histograms also for the various diagnostics on each node.
- I'm working on a dump of CSV for the node-reduced data and per-node, if the user opts in.
- I also implemented "snapshots" which can be set up with varying ranges (across time windows or phase windows). I'm working on finishing up how those get triggered. For now, I'm going to set up two snapshots: (1) the global window; and, (2) a time window that defaults to one second.

PhilMiller commented on src/vt/utils/adt/histogram_approx.h:
```diff
@@ -0,0 +1,573 @@
+      }
+    }
+
+    // append element at the end
+    cs_.resize(cs_.size() + 1);
+
+    // shift over to make room (maybe I should be using a linked list or
+    // \c std::map ?)
+    for (auto i = cs_.size() - 1; i > k; i--) {
+      cs_[i] = cs_[i-1];
+    }
+
+    // insert new centroid here
+    cs_[k] = CentroidType{value, num_times};
```
I think you just want std::vector::insert, rather than doing the resizing, shifting, etc manually. As written, the whole collection potentially gets scanned and copied twice, rather than the once necessary.

lifflander replied:
Addressed in #921

PhilMiller commented on src/vt/utils/adt/histogram_approx.h:
```diff
@@ -0,0 +1,573 @@
+    // insert new centroid here
+    cs_[k] = CentroidType{value, num_times};
+
+    // add \c num_times to total count
+    total_count_ += num_times;
```
Why not put this ahead of the search, to avoid duplication in the exact match case?

lifflander replied:
Addressed in #921

PhilMiller commented on src/vt/utils/adt/histogram_approx.h:
```diff
@@ -0,0 +1,573 @@
+      cs_[j] = cs_[j+1];
+    }
+    // remove the last element, now dead
+    cs_.pop_back();
```
Would it make sense to implement merging insertion once over the size limit a little less naively? Merge appropriate neighbors, then bubble/insertion the new entry into place? So, there's never a need to reallocate or mass-copy.

lifflander replied:
Addressed in #921

PhilMiller commented on src/vt/utils/adt/histogram_approx.h:
```diff
@@ -0,0 +1,573 @@
+    for (auto&& c : in_cs) {
+      add(c->getValue(), c->getCount());
+    }
+  }
```
This doesn't follow the paper's algorithm, and I think loses accuracy as a result. The documented merging algorithm considers nearest-pair bins over the combined set of centroids at once, not one-by-one. For instance, the closest pairs may all exist in `in_hist`, but this approach will never consider that.

lifflander replied:
Addressed in #921

pnstickne commented on src/vt/event/event.cc:
```diff
@@ -298,11 +312,26 @@ void AsyncEvent::testEventsTrigger(int const& num_events) {
   auto& cont = polling_event_container_;
+
+  if (cont.size() > 0) {
+    eventSizeGauge += cont.size();
```
Operators ^_^ I entirely missed it at first..

Anyway, consider eliminating the `> 0` check here and add it into the collector if/as relevant. Recording 0's may be relevant to deal with averages.

lifflander replied:
Ok, I've moved away from operators, removing them.

pnstickne commented on src/vt/event/event.h:
```diff
@@ -203,6 +203,11 @@ struct AsyncEvent : runtime::component::PollableComponent<AsyncEvent> {
 # endif
+
+private:
+  diagnostic::Counter<int64_t> eventPollCount;
```
If standardizing on a size (this would probably be ideal in _most_ cases), can eliminate the use of templates to save some typing and reduce unexpected deviations.

lifflander replied:
I've added a default type for all the diagnostic meters.

pnstickne commented on src/vt/messaging/active.h:
```diff
@@ -1525,6 +1525,38 @@ struct ActiveMessenger : runtime::component::PollableComponent<ActiveMessenger>
+
+private:
+  // Diagnostic counters for sent counts/bytes
+  diagnostic::Counter<int64_t> amSentCount;
```
A comment at the top of block for 'dm' vs 'am' might be useful.

lifflander replied:
I've added that to the comments

pnstickne commented on src/vt/runtime/component/diagnostic_meter.h:
```diff
@@ -0,0 +1,303 @@
+  /**
+   * \brief Convenience method for incrementing---prefix operator
+   */
+  Counter<T>& operator++() {
```
I'd be tempted to nix the operators. Useful, and perhaps too many options.

lifflander replied:
Removed

pnstickne commented:
Cool. I think that'll be a fairly good base going forward.

Obviously not such a fan of the operators - largely that it removes a visual cue on the type of variable/operation..

pnstickne commented on src/vt/messaging/active.cc:
```diff
@@ -74,6 +74,72 @@ ActiveMessenger::ActiveMessenger()
   pushEpoch(term::any_epoch_sentinel);
+
+  // Register counters for AM/DM message sends and number of bytes
+  amSentCount = registerCounter<int64_t>("AM_sent", "active messages sent");
```
The more I see the int64_t, the more I think it could just be elided entirely and used as the only internal data-type.

lifflander replied:
Made that the default

pnstickne commented on src/vt/utils/adt/union.h:
```diff
@@ -338,6 +338,35 @@ struct AllUnique<U, T> {
+  }
+};
+
+template <typename T>
```
Cost/gain benefits?

lifflander replied:
The static type switch is much safer. Guarantees that all switch cases are always handled.

lifflander commented:
So, two PRs have been spun off of this one: https://github.com/DARMA-tasking/vt/pull/923 (type-safe switch on union) and https://github.com/DARMA-tasking/vt/pull/921 (approximate histogram) that have now been merged on develop. I'm rebasing these out of the PR and adjusting the interface for the changes in the type-safe union switch.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/diagnostic_value_format.h  2
- src/vt/runtime/component/diagnostic_meter.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  2
- src/vt/runtime/component/diagnostic_meter.h  2
- src/vt/runtime/component/diagnostic.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

PhilMiller commented on src/vt/runtime/runtime.cc:
```diff
@@ -372,6 +372,7 @@ bool Runtime::initialize(bool const force_now) {
 bool Runtime::finalize(bool const force_now) {
   if (force_now) {
     auto const& is_zero = theContext->getNode() == 0;
+    computeAndPrintDiagnostics();
```
Given how substantial the diagnostics are getting, this may need to default to an output file (akin to `timing.log`) rather than stdout.

lifflander replied:
Yes, this was a hack for just seeing them while I'm working on it. I'm trying to figure out the correct way to configure output.

lifflander replied:
This is done now

PhilMiller commented on src/vt/runtime/component/diagnostic_units.h:
```diff
@@ -0,0 +1,79 @@
+  Bytes,                        /**< Memory units as bytes */
+  Units,                        /**< Units of some entity */
+  UnitsPerSecond,               /**< Units per second */
+  Seconds                       /**< Time units as seconds */
```
Should `BytesPerSecond` be distinguished here, as a specific throughput measurement? Potentially relevant for at least network bandwidth usage and memory allocation.

lifflander replied:
I think it might be worth adding this later. Right now, no metrics are computed in units per section yet.

PhilMiller commented on src/vt/messaging/active.h:
```diff
@@ -1525,6 +1525,38 @@ struct ActiveMessenger : runtime::component::PollableComponent<ActiveMessenger>
+
+  // Diagnostic counters for counting forwarded messages
+  diagnostic::Counter<int64_t> amForwardCount;
+  diagnostic::Gauge<int64_t> amForwardBytesGauge;
```
Given the frequent use of Counter/Gauge combos, I think it would make sense to present a consolidated abstraction. It would probably allow us to do some more meaningful analysis with their data, knowing that they're paired.

lifflander replied:
I've added a new combo

PhilMiller commented on src/vt/runtime/component/diagnostic_value.cc:
```diff
@@ -0,0 +1,117 @@
+    if (update == DiagnosticUpdate::Min) {
+      out->is_valid_value_ = reduced_val.min() != std::numeric_limits<T>::max();
+    } else {
+      out->is_valid_value_ = reduced_val.sum() != 0;
```
Is this a valid test for things that might be decremented, or legitimately have a value of 0?

lifflander replied:
If the sum is 0, that implies that there are no occurrences. It can never be negative. Thus, I disable printing the min/max which will have sentinel values in them.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  2
- src/vt/runtime/component/diagnostic_meter.h  2
- src/vt/runtime/component/diagnostic.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/diagnostic_value_format.h  2
- src/vt/runtime/component/meter/timer.h  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/meter/timer.h  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/meter/timer.h  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/diagnostic_value_format.h  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

pnstickne commented:
@lifflander Bump?

pnstickne commented on cmake/define_build_types.cmake:
```diff
@@ -164,6 +164,15 @@ else()
   set(vt_feature_cmake_zoltan "0")
 endif()
 
+option(vt_diagnostic_enabled "Build VT with diagnostics" ON)
```
Might be "with performance metrics".

lifflander replied:
Done

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/meter/timer.h  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/diagnostic_value_format.h  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/runtime_diagnostics.cc  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented on scripts/run-repeat.sh:
```diff
@@ -0,0 +1,23 @@
+    echo "running to output $file : $i"
+    ##echo mpirun -n $np $pgm "$@" 2>&1
+    i=$(expr $i + 1)
+    mpirun -n $np $pgm "$@" 2>&1 > $file
```
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [Double quote to prevent globbing and word splitting.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander commented on scripts/run-repeat.sh:
```diff
@@ -0,0 +1,23 @@
+    file=$(mktemp)
+    echo "running to output $file : $i"
+    ##echo mpirun -n $np $pgm "$@" 2>&1
+    i=$(expr $i + 1)
```
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/component/meter/timer.h  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/diagnostic_value_format.h  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/meter/timer.h  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/component/meter/gauge.h  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/runtime_diagnostics.cc  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

PhilMiller commented on src/vt/collective/reduce/reduce_scope.h:
```diff
@@ -48,6 +48,8 @@
 #include "vt/utils/adt/union.h"
 
+#include <unordered_map>
+
```
Superfluous, perhaps?

lifflander replied:
This was actually missing from the file so I added it in along the way.

PhilMiller commented on src/vt/runtime/runtime.cc:
```diff
@@ -413,6 +413,13 @@ bool Runtime::finalize(bool const force_now, bool const disable_sig) {
+      computeAndPrintDiagnostics();
+    }
+#   endif
+
```
Can't this go in the diagnostic component's `finalize()`?

lifflander replied:
Not really because this loops through every component, extracts it diagnostics, and then does the global computation over all of them.

PhilMiller commented on src/vt/scheduler/scheduler.cc:
```diff
@@ -67,11 +67,74 @@ namespace vt { namespace sched {
+  using timing::Timing;
+
+  // Triggers to get the in-scheduler-loop time added to diagnostics
+  registerTrigger(
+    BeginSchedulerLoop, [this]{ begin_loop_time_ = Timing::getCurrentTime(); }
+  );
+  registerTrigger(
+    EndSchedulerLoop, [this]{
+      schedLoopTime.update(begin_loop_time_, Timing::getCurrentTime());
+    }
+  );
```
This pattern of starting and stopping a timer is repeated three times just here, and I'm guessing at least a few more by other callers. Would it make sense to bake "start timer X" and "stop time X" into the interface, rather than make the callers keep variables for the start time?

lifflander replied:
Done, added to interface

PhilMiller commented on src/vt/scheduler/scheduler.cc:
```diff
@@ -67,11 +67,74 @@ namespace vt { namespace sched {
+}
+
+void Scheduler::preDiagnostic() {
+  vtLiveTime.update(startup_time_, timing::Timing::getCurrentTime());
```
Assuming `preDiagnostic` might ever be called at any point other than shutdown, this use of `Timer::update` is inconsistent with the scheduler loop and idle time calls. It would seem to repeatedly accumulate time since initialization, rather than just incremental time since the last time diagnostics were computed.

lifflander replied:
Currently, `preDiagnostic` is only called once. When we start calling it multiple times, which might be with snapshots, we can adjust this. There are also some other complexities in doing that so I'm not going too address this now.

PhilMiller commented on src/vt/runtime/component/diagnostic_value.h:
```diff
@@ -0,0 +1,445 @@
+  DiagnosticValueWrapper(ReduceTag, T in_value, bool updated)
+    : value_(in_value),
+      N_(updated ? 1 : 0),
+      min_(updated ? in_value : std::numeric_limits<T>::max()),
+      max_(updated ? in_value : std::numeric_limits<T>::lowest()),
```
Why are these not infinity and negative infinity, respectively?

lifflander replied:
This is actually because `T` might be an integer and `std::numeric_limits<T>::infinity` is zero for integers. This just keeps it simpler I think.

PhilMiller commented on src/vt/runtime/component/diagnostic_value.h:
```diff
@@ -0,0 +1,445 @@
+   */
+  friend DiagnosticValueWrapper<T> operator+(
+    DiagnosticValueWrapper<T> d1, DiagnosticValueWrapper<T> const& d2
+  ) {
```
Are the various formulas here known to be numerically stable? If so, a reference to the description/analysis showing as much. If not, they ought to be. Sums and averages of highly divergent values can have serious issues.

lifflander replied:
Yes, these are numerically stable. I actually got them from Philippe who's an expert in such things. I can ask him for a reference on these.

PhilMiller replied:
So, the *sum* is not numerically stable. Simple addition in arbitrary order can screw things up in a pretty wide variety of ways. This going up a reduction tree helps on average, but doesn't really guard against the bad cases.

PhilMiller commented on src/vt/runtime/component/diagnostic_value.h:
```diff
@@ -0,0 +1,445 @@
+  /**
+   * \internal \brief Get the mean value (use after reduction)
+   *
+   * \return the mean value
```
Should specify that this is intended to be the "arithmetic mean" (as opposed to geometric, harmonic, etc)

lifflander replied:
Fixed

PhilMiller commented on src/vt/runtime/component/diagnostic_value.h:
```diff
@@ -0,0 +1,445 @@
+   *
+   * \return the variance
+   */
+  double var() const { return M2_ * (1.0f / N_); }
```
Do we need to guard against `N_ == 0` here in some fashion? If no, can that be dropped from `avg()`?

lifflander replied:
Fixed

PhilMiller commented on src/vt/runtime/component/diagnostic_value.h:
```diff
@@ -0,0 +1,445 @@
+  /**
+   * \internal \brief Get the computed value (based on update type)
+   *
+   * \note Either returns the mean (when a average update type is applied) or
```
Typo a -> an

lifflander replied:
Fixed

PhilMiller commented on src/vt/runtime/component/diagnostic_value.h:
```diff
@@ -0,0 +1,445 @@
+        v.update(v.getRawValue() + val);
+        break;
+      default:
+        vtAssert(false, "Unknown DiagnosticUpdate---should be unreachable");
```
This should be `vtAbort` I think?

lifflander replied:
Done

PhilMiller commented on src/vt/runtime/component/diagnostic_value.h:
```diff
@@ -0,0 +1,445 @@
+   * \return the value wrapper
+   */
+  DiagnosticValueWrapper<T>& operator[](int snapshot) {
+    vtAssert(
```
This assertion is redundant with the use of `.at`. The error reporting will be only marginally better.

lifflander replied:
Removed

PhilMiller commented on src/vt/runtime/runtime_diagnostics.cc:
```diff
@@ -0,0 +1,397 @@
+    return fmt::format(decimal_format, val);
+  } else {
+    return DF::getValueWithUnits(val, unit, decimal_format, align);
+  }
```
This seems redundant with `apply`. Should one be calling the other? The `default_format/decimal_format` logic is a bit different.

lifflander replied:
These are actually doing different things so I've leaving them as be.

PhilMiller commented on src/vt/runtime/component/diagnostic_value_format.h:
```diff
@@ -0,0 +1,297 @@
+    case DiagnosticUnit::Bytes: {
+      // Start with the largest unit, testing if its appropriate for the value,
+      // and if not, downgrade it until we find one that works
+      auto multiplier = static_cast<int8_t>(MemoryUnitEnum::Petabytes);
```
Out of respect for an extreme but predictable use case, please start this at exabytes or yottabytes - free or total disk space, summed across nodes, when looking at a shared Amazon Web Services Elastic File System. This will typically be in the single-digit exabytes for every such instance.

lifflander replied:
Done, expanded out to Yottabytes

PhilMiller commented on src/vt/messaging/active.h:
```diff
@@ -1553,6 +1553,30 @@ struct ActiveMessenger : runtime::component::PollableComponent<ActiveMessenger>
+
+  // Diagnostic counters for posted irecv counts/bytes
+  diagnostic::CounterGauge amPostedCounterGauge;
+  diagnostic::CounterGauge dmPostedCounterGauge;
```
I'm somewhat confused at this usage of 'Gauge', since such things would generally be taken to measure the evolving level of a variable over time. Using a gauge to accumulate over all time is weird. Think "tidal gauge" vs "water meter"

lifflander replied:
Resolved based on discussion

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/diagnostic_value_format.h  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

PhilMiller commented:
This generally seems to be quite nice. I'm looking forward to having this available, and finding all the many other places in the runtime that are useful to instrument. One that comes to mind is of course memory allocation - the various level trackers are natural gauges, and the cumulative flow of allocations and frees are natural counters/meters

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/diagnostic.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/component/diagnostic_value_format.h  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
Per Phil's suggestion, I need to fix the mean to actually correctly calculate based on the cardinality of each node's counter.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/component/diagnostic.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/diagnostic.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
gcc-7 keeps failing during linking and I don't know why! I reset the cache so the disk space should be larger I think...

lifflander commented:
I think this is ready to merge! `git --check` fails because of the library that was added---can't avoid it

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/diagnostic.impl.h  2
- tests/unit/diagnostics/test_diagnostic_value.cc  4
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/runtime_diagnostics.cc  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/diagnostic.impl.h  2
- tests/unit/diagnostics/test_diagnostic_value.cc  4
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/diagnostic_value_format.h  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/runtime_diagnostics.cc  2
- tests/unit/diagnostics/test_diagnostic_value.cc  4
- src/vt/runtime/component/diagnostic_value_format.h  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/component/diagnostic_value_format.h  5
- tests/unit/diagnostics/test_diagnostic_value.cc  4
- src/vt/runtime/component/diagnostic.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/diagnostic_value_format.h  5
- tests/unit/diagnostics/test_diagnostic_value.cc  4
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/runtime_diagnostics.cc  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- tests/unit/diagnostics/test_diagnostic_value.cc  4
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/diagnostic_value_format.h  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
Here's the azure pipeline build with full system diagnostics enabled to try to find the problem:
https://dev.azure.com/DARMA-tasking/DARMA/_build/results?buildId=10&view=logs&j=3dc8fd7e-4368-5a92-293e-d53cefc8c4b3&t=28db5144-7e5d-5c90-2820-8676d630d9d2

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/diagnostic.impl.h  2
- tests/unit/diagnostics/test_diagnostic_value.cc  4
- src/vt/runtime/component/meter/timer.h  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented:
@PhilMiller Can I get an approve on this?

lifflander commented:
@cz4rs @JacobDomagala @jstrzebonski Can I get a review on this PR? I want to get it approved so I can merge it.

cz4rs commented:
> @cz4rs @JacobDomagala @jstrzebonski Can I get a review on this PR? I want to get it approved so I can merge it.

I'm looking at it right now

cz4rs commented on scripts/run-repeat.sh:
```diff
@@ -0,0 +1,23 @@
+    file=$(mktemp)
+    echo "running to output $file : $i"
+    ##echo mpirun -n $np $pgm "$@" 2>&1
+    i=$(expr $i + 1)
```
https://www.shellcheck.net/ spits out following suggestion (honestly, I don't know which braces are the way to go currently):
```
    i=$(expr $i + 1)
        ^-- SC2003: expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].
 ```

cz4rs replied:
side note: analyzing the scripts could be easily added to CI if we ever want automate it

lifflander replied:
Fixed

cz4rs commented on scripts/run-repeat.sh:
```diff
@@ -0,0 +1,23 @@
+    echo "running to output $file : $i"
+    ##echo mpirun -n $np $pgm "$@" 2>&1
+    i=$(expr $i + 1)
+    mpirun -n $np $pgm "$@" 2>&1 > $file
```
again `shellcheck`, this one seems to be more important actually:
```
    mpirun -n $np $pgm "$@" 2>&1 > $file
              ^-- SC2086: Double quote to prevent globbing and word splitting.
                  ^-- SC2086: Double quote to prevent globbing and word splitting.
                            ^-- SC2069: To redirect stdout+stderr, 2>&1 must be last (or use '{ cmd > file; } 2>&1' to clarify).
                                   ^-- SC2086: Double quote to prevent globbing and word splitting.

Did you mean:
    mpirun -n "$np" "$pgm" "$@" 2>&1 > "$file"
```

lifflander replied:
Fixed

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/vt/runtime/component/meter/gauge.h  1
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/diagnostic_value_format.h  5
- tests/unit/diagnostics/test_diagnostic_value.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented on scripts/run-repeat.sh:
```diff
@@ -0,0 +1,23 @@
+    file=$(mktemp)
+    echo "running to output $file : $i"
+    ##echo mpirun -n $np $pgm "$@" 2>&1
+    i=$(($i + 1))
```
Codacy found an issue: [$/${} is unnecessary on arithmetic variables.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

cz4rs replied:
https://github.com/koalaman/shellcheck/wiki/SC2004

cz4rs reviewed (approved):
Thumbs up after 3 hrs of reading - nothing major comes to mind so far. Considering that it's almost 15k loc, I will be giving it another look tomorrow.

- it looks like it could use extensive manual testing and getting it merged should definitely help in that aspect (to get more hands on it and create the issues as needed)
- unit test for `Runtime::computeAndPrintDiagnostics()` would document the output format and serve as an integration test for the whole thing, although I'm not sure how practical it would be

lifflander commented:
> Thumbs up after 3 hrs of reading - nothing major comes to mind so far. Considering that it's almost 15k loc, I will be giving it another look tomorrow.

Sounds good. By the way, about half of the lines of code are from importing a new optional library (libfort) that prints formatted tables nicely on the terminal.

> 
> * it looks like it could use extensive manual testing and getting it merged should definitely help in that aspect (to get more hands on it and create the issues as needed)
> * unit test for `Runtime::computeAndPrintDiagnostics()` would document the output format and serve as an integration test for the whole thing, although I'm not sure how practical it would be (to write and maintain)

JacobDomagala commented on src/vt/runtime/component/diagnostic.cc:
```diff
@@ -0,0 +1,91 @@
+//@HEADER
+*/
+
+#if !defined INCLUDED_VT_RUNTIME_COMPONENT_DIAGNOSTIC_CC
```
Seems like auto-generated include guard?

lifflander replied:
Removed

JacobDomagala reviewed (approved):
I skimmed through the code, looks good overall.

PhilMiller commented:
I'm back online, and starting another pass of review on this now.

PhilMiller commented on src/vt/runtime/runtime.cc:
```diff
@@ -429,8 +430,15 @@ bool Runtime::finalize(bool const force_now, bool const disable_sig) {
+    }
+#   endif
+
+    auto const& num_units = theTerm->getNumUnits();
```
Not really necessary to change here, but should that printout just be one more diagnostic value, now, rather than something distinguished?

lifflander replied:
Yes, we should change that eventually.

PhilMiller commented on src/vt/runtime/component/diagnostic_value_format.h:
```diff
@@ -0,0 +1,297 @@
+    case DiagnosticUnit::Seconds: {
+
+      // Start with the largest unit, testing if its appropriate for the value,
+      // and if not, downgrade it until we find one that works
```
This comment doesn't describe the reversed sense used for time units as for everything else - smallest unit, and successive upgrades

lifflander replied:
Fixed

PhilMiller reviewed (approved):
Let's get this in.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/runtime_diagnostics.cc  2
- tests/unit/diagnostics/test_diagnostic_value.cc  4
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/component/meter/timer.h  1
- src/vt/runtime/component/meter/gauge.h  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)

lifflander commented on scripts/run-repeat.sh:
```diff
@@ -0,0 +1,23 @@
+    file=$(mktemp)
+    echo "running to output $file : $i"
+    ##echo mpirun -n $np $pgm "$@" 2>&1
+    i=$(($i + 1))
```
Codacy found an issue: [$/${} is unnecessary on arithmetic variables.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander replied:
Codacy found an issue: [$/${} is unnecessary on arithmetic variables.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/vt/runtime/component/meter/timer.h  1
- tests/unit/diagnostics/test_diagnostic_value.cc  4
- src/vt/runtime/component/diagnostic_value_format.h  5
- src/vt/runtime/runtime_diagnostics.cc  2
- src/vt/runtime/component/diagnostic.impl.h  2
- src/vt/runtime/component/meter/gauge.h  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5740129&bid=18776306)
